### DISCUSSION TITLE
How to provide better error messages?

### DISCUSSION THREAD

**Discussion by csells** [petitparser/dart-petitparser#125] [Q&A] [answered]
If I have a parser like so:

```dart
  Parser<String> id() => ((letter() | char('_')) &
          (letter() | digit() | anyOf('_- ()')).star().flatten())
      .map((vals) => vals.listOf<String>().join());
```

and I feed it input like 'foo.1', I'll get an error like this:

```text
"\n" expected at 1:4
```

but the error I want to supply is:

```text
Illegal character '.' found in id.
```

How can I get better error messages in my parsers?

**Comment by renggli** [comment, marked as answer]
The problem in your example is that `foo` is a perfectly valid identifier as part of your specification and the `id` parser succeeds. The error message `"\n" expected at 1:4` is produced somewhere else in your grammar.

A common way to fix this is to use a [negative lookahead](https://github.com/petitparser/dart-petitparser/blob/main/petitparser/lib/src/parser/combinator/not.dart). With it you can assert that your ID is not followed by a dot, i.e.

```dart
Parser<String> id() => (((letter() | char('_')) &
            (letter() | digit() | anyOf('_- ()')).star() &
            char('.').not('end of id expected')))
        .flatten();
```

For the input `foo.1` this gives `end of id expected at 1:4`.

Note: I also moved the `flatten` operator around everything, then you don't need the custom flattening with the `map` operator.
